[ZOOKEEPER-1818] Correctly handle potential inconsistent zxid/electionEpoch and peerEpoch during leader election

lvfangmin opened the pull request:
This is similar to the 3.4 implementation in Jira ZOOKEEPER-1817, main differences with that:

1. removed bcVote, in Vote.equals it will skip compare peerEpoch when one of the version is 0x0.
2. added detailed scenarios which we tried to solve with peerEpoch update and the change in Vote.equals.
3. removed ooePredicate with inlined one, since master code is tracking voteSet..
4. improved the tests to make it cleaner.

lvfangmin commented:
@anmolnar this is the fix for the DONTCARE on trunk, please take a look, I'll port it to 3.5 when it's being reviewed and merged.

anmolnar commented:
Thanks @lvfangmin for working on this.

To be honest I'm lost following how FLE works and additionally what issue is this patch intended to fix. I only ran a few blind test with manually restarting ZK instances in a 3-node ensemble and haven't seen any issue with joining so far.

I'd like to do some more stress testing on this and get an understanding on how FLE works before committing. Hopefully early next week. First of all, what's the difference between current and accepted epoch and how election round is related?

lvfangmin commented:
@anmolnar following are my understanding about the acceptedEpoch, currentEpoch and electionEpoch:

* acceptedEpoch : the previous epoch we accepted so far, usually is the epoch is the highest zxid on that server.
* currentEpoch  : the current epoch after syncing with the new leader, it's based on the maximum acceptedEpoch in the quorum, and usually it's the max(acceptedEpoch) + 1. The currentEpoch is used as the peerEpoch in the leader election, as we know (sid, zxid, peerEpoch) are the set used to decide a leader.
* electionEpoch : not part of the factors to decide leader, but it's used as a logical clock to avoid considering a vote delayed from a while ago.

Basically, we know there is a corner case where the learner may not update it's zxid, peerEpoch, and electionEpoch after leader election (check the new comment I added Leader.updateElectionVote), peerEpoch is fixed with a hack solution, but we cannot easily update the zxid and electionEpoch, so we try to ignore it. But IGNOREVALUE introduced will have compatible issue when rolling upgrade ensemble, that's why we introduced version in notification, and only compare id or peerEpoch based on version.

anmolnar commented:
@lvfangmin Thanks for the clarification!
Still testing your patch, should be done by tomorrow.

lvfangmin commented:
May need an extra +1 from another committer, @hanm do you have time to review this?

hanm commented:
@lvfangmin will review and provide feedback tonight. thanks

hanm commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/FastLeaderElection.java:
```diff
@@ -1023,9 +1016,9 @@ else if (validVoter(n.sid) && validVoter(n.leader)) {
-
                                 Vote endVote = new Vote(proposedLeader,
-                                        proposedZxid, proposedEpoch);
+                                        proposedZxid, logicalclock.get(), 
```
nit: indentations looks a little bit off here?

lvfangmin replied:
@hanm for the general indention after { we use 4 white spaces, but for the statement breaking we usually use 8 white spaces, isn't that true for ZK codebase, I saw the previous code is using this rule.

Let me know if this is not true, I can update those.

hanm commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/FastLeaderElection.java:
```diff
@@ -1042,14 +1035,13 @@ else if (validVoter(n.sid) && validVoter(n.leader)) {
-
-                                Vote endVote = new Vote(n.leader, n.zxid, n.peerEpoch);
+                                Vote endVote = new Vote(n.leader, 
+                                        n.zxid, n.electionEpoch, n.peerEpoch);
```
nit: indentation

hanm commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/FastLeaderElection.java:
```diff
@@ -1058,29 +1050,20 @@ else if (validVoter(n.sid) && validVoter(n.leader)) {
                             }
-                            Vote endVote = new Vote(n.leader, n.zxid, n.peerEpoch);
+                            Vote endVote = new Vote(n.leader, n.zxid, 
+                                    n.electionEpoch, n.peerEpoch);
```
nit: indentation

hanm commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumPeer.java:
```diff
@@ -1989,6 +1989,38 @@ private boolean updateVote(long designatedLeader, long zxid){
+     *    join, because z2 is reporting peer epoch 0xb8, while z3 is reporting 
+     *    0xb9.
+     *
+     * By updating the election vote after actually following leader, we can 
```
should be 'election epoch' or 'electionEpoch' rather than 'election vote'.

lvfangmin replied:
It's align with the function name which is updating the vote, although we only updated the electionEpoch here.

hanm commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumPeer.java:
```diff
@@ -1989,6 +1989,38 @@ private boolean updateVote(long designatedLeader, long zxid){
+     * around by skipping comparing the zxid and electionEpoch when counting for 
+     * votes for out of election servers during looking for leader.
      * 
      * @see https://issues.apache.org/jira/browse/ZOOKEEPER-1732
```
would this be better syntax wise: {@see https://issues.apache.org/jira/browse/ZOOKEEPER-1732}

hanm commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumPeer.java:
```diff
@@ -1989,6 +1989,38 @@ private boolean updateVote(long designatedLeader, long zxid){
      * Updates leader election info to avoid inconsistencies when
      * a new server tries to join the ensemble.
+     *
+     * Here is the inconsistency scenario we try to solve by updating the peer 
```
awesome comments.

hanm commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Vote.java:
```diff
@@ -125,11 +125,47 @@ public boolean equals(Object o) {
+             *    And given that only one ensemble can be running at a single point 
+             *    in time and that each epoch is used only once, using only id and 
+             *    epoch to compare the votes is sufficient.
+             *
```
maybe move the reference of {@see https://issues.apache.org/jira/browse/ZOOKEEPER-1805} from bottom to here, save some brain cycles for pattern matching between comments and issues.

hanm commented on zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/FLEOutOfElectionTest.java:
```diff
@@ -0,0 +1,136 @@
+        // Set the logical clock to 1 on fle instance of server 3.
+        fle.logicalclock.set(0x1);
+
+        Assert.assertTrue("Termination predicate failed", 
```
I don't think we are testing any predicates here - maybe this was from the old patch / 3.4 where the `termPredicate` was still active. Suggest to rephrase "Termination predicate failed" to something else.

lvfangmin replied:
Forgot to make this change last time, will update.

lvfangmin commented:
@hanm is the latest change looking good to you?

anmolnar commented:
Committed to master branch. Thanks @lvfangmin !
Please create another pull request for branch-3.5.

lvfangmin commented:
Thanks @anmolnar, I'll send out the PR for 3.5.

lvfangmin commented:
Here is the PR for 3.5: #714, thanks @mkedwards.
